docs/newtypes/index.md

```diff
@@ -0,0 +1,310 @@
+
+def lookup(accountNumber: AccountNumber, sequenceNumber: SequenceNumber) =
+  ???
+
+lookup(accountNumber, sequenceNumber)    // okay
+// lookup(sequenceNumber, accountNumber) // does not compile
+```
+
+This is great but how do we construct `AccountNumber` and `SequenceNumber` values and how do we access the underlying `Int` values to work with them?
+
+## Constructing New Types
+
+The easiest way to construct an instance of the new type is to use the `apply` method on the new type object.
+
+```scala mdoc
+val accountNumber: AccountNumber =
+  AccountNumber(1)
+```
+
+You can think of the new type object as being the "companion object" of the new type and it comes included with an `apply` method to make it easy for us to construct instances of the new type.
+
+If we want to be more explicit we can also always use the `wrap` operator on the new type object to "wrap" a value of the underlying type in the new type. For example:
+
+```scala mdoc
+val sequenceNumber: SequenceNumber =
+  SequenceNumber.wrap(2)
+```
+
+If you want to construct a collection of values of the new type from values of the underlying type you can use the `wrapAll` operator.
+
+```scala mdoc
+import zio.Chunk
+
+val accountNumbers: Chunk[AccountNumber] =
+  AccountNumber.wrapAll(Chunk(3, 4, 5))
+```
+
+The `wrapAll` operator is extremely efficient because it does not have to traverse the collection at all. Internally we know that `AccountNumber` and `Int` are the same type so we can freely convert a `Chunk[Int]` to a `Chunk[AccountNumber]` without having to traverse the collection.
+
+We can also define our own more specialized constructors for our new type in terms of `apply`, `wrap`, and `wrapAll`. The new type object provides a convenient place for us to do this.
+
+For example, we could define a simple constructor for an initial sequence number like this:
+
+```scala mdoc:reset-object
+import zio.prelude.Newtype
+
+object SequenceNumber extends Newtype[Int] {
+  val initial: SequenceNumber =
+    SequenceNumber(0)
+}
+type SequenceNumber = SequenceNumber.Type
+
+val initial: SequenceNumber =
+  SequenceNumber.initial
+```
+
+This let's us construct instances of the new type in a very ergonomic way, where constructors for the new type are on the new type companion object just like constructors for any other type would be.
+
+## Deconstructing New Types
+
+In addition to constructing instances of the new type we want to be able to access the value underlying the new type to perform operations on it.
+
+We can do this using the `unwrap` operator on the new type companion object.
+
+```scala mdoc
+val zero: Int =
+  SequenceNumber.unwrap(initial)
+```
+
+We can also use the `unwrapAll` operator to convert a collection of values of the new type into the underlying type.
+
+```scala mdoc
+import zio.Chunk
+
+val sequenceNumbers: Chunk[SequenceNumber] =
+  SequenceNumber.wrapAll(Chunk(1, 2, 3))
+
+val ints: Chunk[Int] =
+  SequenceNumber.unwrapAll(sequenceNumbers)
+```
```

Comment:
you could reiterate the point of no runtime overhead for deconstructing as well though the reader will probably assume it anyhow.

Reply:
Good idea! Done now.